rust/src/dns/dns.rs

```diff
@@ -769,8 +770,10 @@ pub unsafe extern "C" fn rs_dns_parse_request(
     flow: *const core::Flow, state: *mut std::os::raw::c_void, _pstate: *mut std::os::raw::c_void,
     stream_slice: StreamSlice, _data: *const std::os::raw::c_void,
 ) -> AppLayerResult {
-    let state = cast_pointer!(state, DNSState);
-    state.parse_request_udp(flow, stream_slice);
+    let _ = catch_unwind(|| {
+        let state = cast_pointer!(state, DNSState);
+        state.parse_request_udp(flow, stream_slice);
+    });
```

Comment:
I'm trying to understand why we're using it here. I'm reading `Don't build your programs to unwind under normal circumstances. Ideally, you should only panic for programming errors or extreme problems.` at https://doc.rust-lang.org/nomicon/unwinding.html where I also see `You must absolutely catch any panics at the FFI boundary!` so I kind of see why we want to do this. But,
1. What points of panic/undefined behavior can happen on the enclosed code?
2. How do we know that the panic caused here for whatever reason would be implemented with unwinding? Or, do we not care about it as it is not incurring any runtime cost in either case?

Thank you for this! :)

Reply:
For reference https://redmine.openinfosecfoundation.org/issues/3333

Reply:
> I'm trying to understand why we're using it here. I'm reading `Don't build your programs to unwind under normal circumstances. Ideally, you should only panic for programming errors or extreme problems.` at https://doc.rust-lang.org/nomicon/unwinding.html where I also see `You must absolutely catch any panics at the FFI boundary!` so I kind of see why we want to do this. But,
> 
> 1. What points of panic/undefined behavior can happen on the enclosed code?

Simple ones that we can catch would be programming errors like walking off the end of an array. In C this might create a segfault, but is a catchable error in Rust.  These warnings are more to prevent people from using unwinding as a flow control mechanism, or treating them like exceptions when coming from C++, Python, etc.

I also disagree about the *must* on FFI boundaries.  Currently, we crash if the Rust code panics at the FFI boundary, and that's a reasonable outcome.

> 2. How do we know that the panic caused here for whatever reason would be implemented with unwinding? Or, do we not care about it as it is not incurring any runtime cost in either case?

We don't care.  If the panic is unwindable we'll catch it and not crash.  If it's not catchable, perhaps a deep-rooted segfault, or an illegal instruction the program should abort and crash.

All the Rust web frameworks like Axum, Rocket, Warp, implement such unwinding in their request handlers, so one bad request handler won't take down the server, this is a very similar approach.

Reply:
Thank you for the explanation!
Do I get it right that if there is actually an error that can be unwinded, it'll affect the performance negatively?
Do we log it in stats somehow to indicate what's causing the possible slowdown?

Reply:
> Do I get it right that if there is actually an error that can be unwinded, it'll affect the performance negatively?

Yes, it should a little as it has to do some resetting of data. Plus that line that gets log that we can't control is slow.  All prints to the console are slow though.

The alternative is crashing.